feat(api): add protocol engine setTipState command

jbleon95 opened the pull request:
# Overview

This PR is a continuation of #19363.

This PR adds a new protocol engine command called `setTipState`, which given a labware ID, list of well names and a `TipRackWellState`, will set all the tip rack wells for that tip rack to the requested state in the `TipStore`, after verifying the labware ID is associated with a tip rack and that all the wells given exist for that tip rack.

This command was then used to replace the `reset_tips` action that was previously being used by `Labware.reset()` to set all the `TipRackWellState`s of a tip rack back to `CLEAN`. While this does add an extra protocol engine command to older protocols where they didn't exist before, this creates no observable difference in the protocol, allows us to track when this is called via looking at the command history, and allows use to remove the `reset_tips` related code.

Additionally the PR adds a new style unit test for `LabwareStore` and `LabwareView` that tests both together.

## Test Plan and Hands on Testing

Tested this protocol passes analysis, as it would fail if the `reset()` function no longer worked properly.

```python
requirements = {
    "robotType": "Flex",
    "apiLevel": "2.25"
}

metadata = {
    "protocolName":'Reset tips test',
}

def run(protocol_context):
    trash = protocol_context.load_trash_bin('A3')
    tiprack = protocol_context.load_labware("opentrons_flex_96_tiprack_50ul", "C2")
    pipette = protocol_context.load_instrument("flex_1channel_50", "left", tip_racks=[tiprack])

    for _ in range(96):
        pipette.pick_up_tip()
        pipette.drop_tip()

    tiprack.reset()

    for _ in range(96):
        pipette.pick_up_tip()
        pipette.drop_tip()
```

## Changelog

- added a `setTipState` command to protocol engine
- removed `ResetTipsAction` and associated methods/action handling and replaced it with `setTipState`

## Review requests

Open to changing the name of the command and parameters if someone has a better suggestion.

## Risk assessment

Medium, this will slightly change the commands generated by older protocols but ultimately for reasons described above this should be an improvement.

jbleon95 committed:
added set tip state PE command

jbleon95 committed:
replace reset_tips engine client with an engine command

jbleon95 committed:
remove reset tips action and related sync client and engine commands

jbleon95 committed:
add PE unit test

jbleon95 committed:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pe_set_tip_state_command

jbleon95 committed:
do some checking of labware and wells provided and new labware state PE test file

SyntaxColoring commented on api/tests/opentrons/protocol_engine/state/test_labware_state.py:
```diff
@@ -0,0 +1,153 @@
+        subject_view.raise_if_wells_are_invalid(
+            "labware-id", ["well-1, well-2", "well-4"]
+        )
+        subject_view.raise_if_wells_are_invalid("labware-id", ["well-5"])
```
I think we want this to be three separate `with pytest.raises(...)` blocks, right? In order to test that each of the `raise_if_wells_are_invalid()` calls raises, as opposed to testing that any of the `raise_if_wells_are_invalid()` calls raises.

SyntaxColoring reviewed (approved):
Awesome, thank you, LGTM!

jbleon95 committed:
unit test fix per review

sfoster1 reviewed (changes requested):
Please add typescript types and at least a provisional command text for this in the frontend so we don't forget! Happy to work with you to do this.

sfoster1 commented on api/src/opentrons/protocol_engine/actions/actions.py:
```diff
@@ -314,7 +307,6 @@ class SetErrorRecoveryPolicyAction:
     SetDeckConfigurationAction,
     AddAddressableAreaAction,
     AddLiquidAction,
-    ResetTipsAction,
```
Have you tried pushing this pr to a robot that has a past protocol with a `ResetTipsAction` in it? does it cause protocol run rehydration errors to not have this exist anymore?

SyntaxColoring replied:
It's a good idea to test, but I expect us to be good. This deletes an `opentrons.protocol_engine` action, which aren't modeled as JSON nor stored in the database; unlike `robot_server` actions, which are.

jbleon95 replied:
Tested this on the robot and saw no issues on the ODD or looking at protocol history for an old protocol that used the action

sfoster1 commented on shared-data/command/schemas/14.json:
```diff
@@ -5138,6 +5138,58 @@
       "title": "SetTargetLidTemperatureParams",
       "type": "object"
     },
+    "SetTipStateCreate": {
```
should be a change in schema 15 I believe for edge

sfoster1 reviewed (changes requested):
couple other little things - i think we're on schema 15 in edge after the concurrent module actions stuff, and I'm a bit worried about persistence directory shenanigans when deleting an action

jbleon95 committed:
Merge branch 'edge' of https://github.com/Opentrons/opentrons into pe_set_tip_state_command

jbleon95 committed:
update schema 15 not 14

jbleon95 committed:
command text for the new set tip state command

jerader commented on shared-data/command/types/setup.ts:
```diff
@@ -337,3 +350,9 @@ interface LoadLidResult {
+export interface SetTipStateParams {
+  labwareId: string
+  wellNames: string[]
+  tipWellState?: 'clean' | 'used' | 'empty'
```
small nit: i recommend turning this into a type that is exportable, in case another place needs to import it.

```suggestion
  tipWellState?: TipWellState
```

then up above 

```
export type TipWellState = 'clean' | 'used' | 'empty'
```

jerader reviewed (commented):
run log command stuff looks good! just left a nit. Also, would be great if you can make a new unit test file and write a test 😄

jbleon95 committed:
nit fix and unit test

jerader reviewed (approved):
command text code lgtm, nice!

sfoster1 reviewed (approved):
Looks good to me, thank you!

One thing I realized that might be nice is to put the full tip state of every well in the labware in the command result; that way, if the app wants to render the tiprack as-of the end of the command, they don't have to accumulate previous commands.

github-actions[bot] committed:
fix(analyses-snapshot-testing): heal pe_set_tip_state_command snapshots (#19502)

This PR was requested on the PR
https://github.com/Opentrons/opentrons/pull/19493

Co-authored-by: jbleon95 <12025228+jbleon95@users.noreply.github.com>
